### ISSUE TITLE
Predicate incorrectly pushed down through UNION ALL with window functions

### ISSUE THREAD

**Issue by hlinnaka** [greenplum-db/gpdb#5762]
### Greenplum version or build

Master, 5X_STABLE

### Expected behavior

```
select * from (
  select i, row_number() over (partition by i/3) from generate_series(1, 5) i
  UNION ALL
  select i, row_number() over (partition by i/3) from generate_series(1, 5) i
) as sq WHERE i = 3;
 i | row_number 
---+------------
 3 |          3
 3 |          3
(2 rows)
```

### Actual behavior

```
 i | row_number 
---+------------
 3 |          1
 3 |          1
(2 rows)
```

The planner incorrectly pushes down the `i = 3` qual through the UNION ALL, and below the window function evaluation. So the row_number() window function only sees the row with `i = 3`, which is wrong. The qual is supposed to be evaluated after the window function.

**Comment by guofengrichard** [comment]
Interesting issue. I'd like to take it.

**Comment by guofengrichard** [comment]
This issue was introduced in this commit: https://github.com/greenplum-db/gpdb/commit/a802eb2f477aebbcefb93902347ebfddc2ae3bfc

**Comment by hlinnaka** [comment]
This was fixed on master, in the 9.3 merge. Perhaps by upstream commit 964c0d0f80e485dd3a4073e073ddfd9bfdda90b2, although I haven't tested it.

**Comment by berlin-ab** [comment]
Is this waiting on a fix from master to be backported down?
